tests/python/contrib/test_ethosu/test_codegen.py

```diff
@@ -48,122 +46,147 @@ def get_shape_expr(in_expr, out_expr):
+        converter.representative_dataset = representative_dataset
+        converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
+        converter.inference_input_type = tf.int8
+        converter.inference_output_type = tf.int8
+        tflite_model = converter.convert()
+        return tflite_model
+
+    def create_tflite_graph_double():
+        class Model(tf.Module):
+            @tf.function
+            def tf_function_double(self, x):
+                # Use tf.nn API to create the model with two convolutions
+                op = tf.nn.conv2d(
+                    x,
+                    filters=tf.constant(np.random.uniform(size=kernel_shape), dtype=tf.float32),
+                    strides=(1, 1),
+                    padding=padding,
+                    data_format="NHWC",
+                    dilations=1,
+                )
+                # Second convolution
+                op2 = tf.nn.conv2d(
+                    op,
+                    filters=tf.constant(np.random.uniform(size=kernel_shape), dtype=tf.float32),
+                    strides=(1, 1),
+                    padding=padding,
+                    data_format="NHWC",
+                    dilations=2,
+                )
+                return op2
+
+        model = Model()
+        concrete_func = model.tf_function_double.get_concrete_function(
+            tf.TensorSpec(ifm_shape, dtype=tf.float32)
         )
 
-        # Assumes only two runtime.Modules are created -- i.e. single offload module
-        imported_modules = compiled_models[0].executor_factory.lib.imported_modules
-        assert len(imported_modules) == 2
-        ethosu_module = imported_modules[0]
-
-        # Verify generated C source
-        get_cs = tvm._ffi.get_global_func("runtime.module.ethosu.getcs")
-        cmms = get_cs(ethosu_module)
-        cmms = bytes.fromhex(cmms)
-        infra.print_payload(cmms)
-        infra.verify_source(compiled_models, accel_type)
+        # Convert the model
+        def representative_dataset():
+            for _ in range(100):
+                data = np.random.rand(*tuple(ifm_shape))
+                yield [data.astype(np.float32)]
+
+        converter = tf.lite.TFLiteConverter.from_concrete_functions([concrete_func])
+        converter.optimizations = [tf.lite.Optimize.DEFAULT]
+        converter.representative_dataset = representative_dataset
+        converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
+        converter.inference_input_type = tf.int8
+        converter.inference_output_type = tf.int8
+        tflite_model = converter.convert()
+        return tflite_model
+
+    tflite_graph_single = create_tflite_graph_single()
+    tflite_model_single = tflite.Model.Model.GetRootAsModel(tflite_graph_single, 0)
+
+    tflite_graph_double = create_tflite_graph_double()
+    tflite_model_double = tflite.Model.Model.GetRootAsModel(tflite_graph_double, 0)
```

Comment:
I think tflite_model_double and tflite_model_single beyond this point has the same code, if I am not mistaken. Could we refactor that bit re-use the same code ?

Reply:
ack

Reply:
Wouldn't it make more sense to have single and double graphs as separate tests with their own parameters? Currently, if that test fails, it is not immediately obvious whether it was the single or double graph test that failed

Reply:
Agreed. That's much better